src/worker/mod.rs

```diff
@@ -369,51 +412,86 @@ impl<
                 .await?;
         }
 
-        let exit = self.listen_for_heartbeats(&statuses).await;
+        // Either we have already taken this future and are polling it elsewhere (see Worker::run_to_completion),
+        // or they've never provided it to us (reminder: they do so via WorkerBuilder::with_graceful_shutdown)
+        let cancel_signal = self.shutdown_signal.take();
```

Comment:
I'm worried about taking the shutdown signal here, because it means that if the user decides to reconnect using the same `Worker`, their graceful shutdown future is no longer there and active. I think we should only `take` it if it resolves (because then it's _definitely_ not useful again in the future). And then we should also document (on `with_graceful_shutdown`) that if the future resolves, it is then removed and no longer taken into account if the `Worker` _is_ reused.

Reply:
You are right, there is one scenario that I've missed: if we get an error from heartbeat(), the worker will _not_ be marked as `terminated` and so they may try to call `run` again. In all other cases the worker is marked as terminated and running this work will panic at the check we are performing. So in all other cases they will need to construct a new worker and run it again.

As for taking in case the future resolves, I _think_ there is no need to do so, since if it resolves we are getting into the block marking the worker as terminated. Added to the docs for `WorkerBuilder::with_graceful_shutdown` and `Worker::run`

Also updated this method to not take the signal.

Reply:
Hmm, but now the code has an `Arc<Mutex` instead, which feels unnecessary. I _think_ you should be able to keep it just `ShutdownSignal`. The part that's making your life difficult (I think) is the `Option`, because it's what making you introduce an additional `async` block there. You could perhaps change `ShutdownSignal` to _not_ be an `Option`, and instead set it to [`std::future::pending`](https://doc.rust-lang.org/std/future/fn.pending.html) when `None`. That way your `select!` arm can simply be:

```rust
_ = &mut self.shutdown_signal => {}
```

and that arm will simply never resolve if no shutdown signal was given, and you don't need an `Option` _or_ an `Arc<Mutex`.

Reply:
Actually option never bothered me a lot, it is rather the _mutable self_ that we already need for the heartbeat task and  so we cannot we give it to any other arm. You can image now why I was suggesting to take the signal initially (but I really missed that one scenario where the worker will be re-used), and why I am suggesting to put it behind a tokio mutex now. I think it is fine in this case where we _know_ there is only one arm wanting this lock and so there will be no performance issues here.

As for the `std::future::pending`, I believe it is an elegant construct, but I am not sure the shutdown signal arm should be polled at all if they've never provided the signal. I am not sure that the runtime can inspect this future and see that it will _never_ resolve and so it should not waste _any_ efforts on it. Plus we are spawning workers (`Worker::for_worker`) and will need to pass `Box::pin(std::future::pending())` for those "processing" workers that actually do not need the signalling future (only the coordinator does).

Reply:
Ah, and that's because Rust doesn't realize that `self.listen_for_heartbeats` doesn't actually need to mutable borrow all the _fields_ of `self`. Which is what the `Option` was trying to get around. Okay, I'm with you now. In that case, I think we should rather bring back the `Option` than have an unnecessary `Arc<Mutex`. It's not so much about the performance as it is about the cognitive complexity — we now need to remember why it is that thing has to be reference counted _and_ locked, and we now introduce the potential for deadlocks if we later start also taking it somewhere else.

A permanently-pending arm should indeed never cause any waste — that future will never have its waker awoken, so it will never be polled again. Wouldn't worry on that part about using `std::future::pending`.

Reply:
yeah I wish there were a "sharded" mutable reference and I guess the compiler will support it the future.

As for the option and the mutex, I am not sure those are related. The option actually never helped in this situation (reminder: my taking a signal from an option and not putting it back was a _mistake_), it is the counted async mutex that did. As for the deadlocks - as long as `Worker::run` takes a _mutable_ reference - we are ok. I cannot imagine how they/we manage to cause a dead lock.

wdyt?

Reply:
I figured it was easier to show than try to explain; here's what I meant: https://github.com/jonhoo/faktory-rs/pull/57/commits/8de6397a497e6bab43ae522ea1490582c032b289 (pushed directly to this branch).

If you're happy with it, I think we're good to merge!

Reply:
Oh boy I could have simply put it back :facepalm:

Reply:
honestly I never considered it since I was sure I've moved and there is no way I can get it in the other arm

Reply:
oh.. and _now_ I see how the async block in your earlier comment is related :)

Reply:
One more tweak: https://github.com/jonhoo/faktory-rs/pull/57/commits/6eb5b1726af3ac0a20861a29f51a0fb75f79e5b5

And yeah, that's why I need the `&mut shutdown_signal` in the `select!` arm — so that the arm doesn't _move_ `shutdown_signal`, it only borrows it. Once you _enter_ an arm in a `select!`, all the other arms are cancelled, and no longer borrow whatever they were borrowing.

Reply:
Also note that while there is an `Option` there now, it's used somewhat differently than in your original. It's no longer "none == no shutdown signaler"; `None` is simply used as a temporary holding value.